src/pytest_plugins/filler/gen_test_doc/page_props.py

```diff
@@ -0,0 +1,276 @@
+from dataclasses import asdict, dataclass
+from pathlib import Path
+from typing import Any, Dict, List, Optional
+
+import mkdocs_gen_files  # type: ignore
+from jinja2 import Environment
+
+from ethereum_test_tools import Opcodes
+
+
+def apply_name_filters(input_string: str):
+    """
+    Apply a list of capitalizations/regexes to names used in titles & nav menus.
+
+    Note: As of PragueEIP7692 With 634 doc pages, this function constitutes ~2.0s
+    of the total runtime (~5.5s). This seems to be insignificant with the time
+    taken by mkdocstrings to include the docstrings in the final output (which)
+    is a separate mkdocs "build-step" occurs outside the scope of this plugin.
+    """
+    word_replacements = {
+        "acl": "ACL",
+        "bls 12": "BLS12",
+        "bls12 g1add": "BLS12_G1ADD",
+        "bls12 g1msm": "BLS12_G1MSM",
+        "bls12 g1mul": "BLS12_G1MUL",
+        "bls12 g2add": "BLS12_G2ADD",
+        "bls12 g2msm": "BLS12_G2MSM",
+        "bls12 g2mul": "BLS12_G2MUL",
+        "bls12 map fp2 to g2": "BLS12_MAP_FP2_TO_G2",
+        "bls12 map fp to g1": "BLS12_MAP_FP_TO_G1",
+        "bls12 pairing": "BLS12_PAIRING_CHECK",
+        "eips": "EIPs",
+        "eof": "EOF",
+        "vm": "VM",
+    }
+    # adding these is the expensive part
+    opcode_replacements = {str(opcode): str(opcode) for opcode in Opcodes if str(opcode) != "GAS"}
+    all_replacements = {**word_replacements, **opcode_replacements}
+    for word, replacement in all_replacements.items():
+        input_string = re.sub(rf"(?i)\b{re.escape(word)}\b", replacement, input_string)
+
+    regex_patterns = [
+        (r"eip-?([1-9]{1,5})", r"EIP-\1"),  # Matches "eip-123" or "eip123"
+    ]
+    for pattern, replacement in regex_patterns:
+        input_string = re.sub(pattern, replacement, input_string, flags=re.IGNORECASE)
+
+    return input_string
+
+
+def snake_to_capitalize(string: str) -> str:  # noqa: D103
+    """
+    Converts valid identifiers to a capitalized string, otherwise leave as-is.
+    """
+    if string.isidentifier():
+        return " ".join(word.capitalize() for word in string.split("_"))
+    return string
+
+
+def sanitize_string_title(string: str) -> str:
+    """
+    Sanitize a string to be used as a title.
+    """
+    return apply_name_filters(snake_to_capitalize(string))
+
+
+def nav_path_to_sanitized_str_tuple(nav_path: Path) -> tuple:
+    """
+    Convert a nav path to a tuple of sanitized strings for use in mkdocs navigation.
+    """
+    return tuple(sanitize_string_title(part) for part in nav_path.parts)
+
+
+@dataclass
+class PagePropsBase:
+    """
+    Common test reference doc page properties and definitions.
+
+    The dataclass attributes are made directly available in the jinja2
+    found in `docs/templates/*.j2`.
+    """
+
+    title: str
+    source_code_url: str
+    valid_from_fork: str
+    path: Path
+    pytest_node_id: str
+    package_name: str
+
+    def get_template(self) -> str:
```

Comment:
```suggestion
    @property
    @abstractmethod
    def template(self) -> str:
      pass
```

And then define the attribute in each subclass:
```python
@dataclass
class FunctionPageProps(PagePropsBase):
  ...
  template: str = "function.md.j2"

Reply:
Thanks, that's much nicer!